### ISSUE TITLE
What does modes do in alacritty keybindings

### ISSUE THREAD

**Issue by blueray453** [alacritty/alacritty#4110]
The config file says:

```
# Values for `mode`:
#   - ~AppCursor
#   - AppCursor
#   - ~AppKeypad
#   - AppKeypad
```

However, i could not find what these modes do. Or, what are they used for. Can you please describe what purpose they serve.

**Comment by chrisduerr** [comment]
These are modes the terminal emulator can be set to. AppKeypad changes the way that certain characters are received for example, so this mode is used to handle that. Other modes like `Alt` is for the alternate screen buffer for example, which is used when you're inside applications like vim.

But it looks like your configuration file might be a bit out of date if that's how it looks to you, I'd recommend looking at the latest one. That should explain at least a little bit why these exist, even though it doesn't state what each one of them does.

**Comment by blueray453** [comment]
Sorry, I found the latest one

```
#    - AppCursor
#    - AppKeypad
#    - Alt
#    - Vi
#
#    A `~` operator can be used before a mode to apply the binding whenever
#    the mode is *not* active, e.g. `~Alt`.
```
I could not understand

> AppKeypad changes the way that certain characters are received for example, so this mode is used to handle that.

Can you please explain what the following lines mean in plain english.
```
#- { key: PageUp,    mods: Shift,   mode: AppCursor, action: ScrollPageUp,  }
#- { key: PageUp,    mods: Shift,   mode: AppKeypad, action: ScrollPageUp,  }
```

I made up the last two lines just to understand the meaning of `AppCursor` and `AppKeypad`

**Comment by chrisduerr** [comment]
There's really no simpler way to put this. If you're in AppKeypad/AppCursor mode, the binding is active. That said, if you don't understand it, you're free to not use mode specific bindings.

**Comment by blueray453** [comment]
I actually don't understand if a feature is not explained or documented then how anyone will be able to use it. However, it is a open source project and you are probably working out of passion. Please keep doing the great work.

**Comment by kchibisov** [comment]
>I actually don't understand if a feature is not explained or documented then how anyone will be able to use it.

If you don't really understand what those modes are for you likely shouldn't care about them much. Those modes are usually activated by apps like `less`.`vim`,`tmux`, `emacs` so the should change the way terminal handles certain key bindings. That's true for `AppKeypad`,`AppCursor`, and  `Alt` modes, since those are set by applications, unlike `Vi` mode, which is purely alacritty thing. Also having bindings for such modes is usually necessary, since you having `PageUp`/`PageDown` and other scrolling actions in `Alt` mode doesn't make any sense, since there's no scrollback at all, and it's being handled by the app you're using.

The `AppCursor`/`AppKeyboard` are similar modes, in which terminal changes the way it sends/handles certain things. Those 3 modes are not alacritty only thing, it's true for every terminal(that implements them ofc), the names `AppCursor` and `AppKeypad`
are coming from xterm, if that matters.

`AppCursor` is triggered by `printf "\e[?1h\a"`
`Alt` is triggered by `printf "\a[?1049h\a"`
`AppKeypad` is triggered by `printf "\e="`

**Comment by blueray453** [comment]
thank you very much. I needed to understand those because of paranoia. 

I feel uncomfortable as long as I do not understand something I need or am using.

**Comment by bezumec82** [comment]
Sorry, but this is not clear.
Do you want to say, that key bindings will be changed when some application is started?
How does it achieved - by parsing the terminal command?
So if I use Vim instead of Vi, I must write Vim mode?
Cause in the present state of configs I can't bind Ctrl+Shift+V to block selection in Vim and Ctrl+V to Paste in terminal.

`key_bindings:
  - { key: V,              mods: Control,        action: Paste            }
  - { key: C,              mods: Control,        action: Copy             }
  - { key: C,              mods: Control|Shift,  chars: "\x03"            }
  - { key: V,                           mode: Vi|~Search, action: ToggleNormalSelection   }
  - { key: V,      mods: Shift,         mode: Vi|~Search, action: ToggleLineSelection     }
  - { key: V,      mods: Control|Shift, mode: Vi|~Search, action: ToggleBlockSelection    }
  - { key: V,      mods: Alt,           mode: Vi|~Search, action: ToggleSemanticSelection }
`

**Comment by bezumec82** [comment]
ok, I tried several tricks - obviously, the mode is a dead feature.
Bad code is a bad code - this is not problem with user.
I'm low-level programmer with 15 years of experience - I know what mode is.
I will look source later and leave some comments.
B.R.

**Comment by bezumec82** [comment]
The problem at this moment looks like V+Ctrl and V+Ctrl+Shift are handled in the same way.
I believe that this is not a problem with parsing Control|Shift and just Control. This is problem with event handling - I think you have some kind of Chain-of-Responsibility handling technique, and you catch event too early.
